text/0000-custom-model-classes.md

```diff
@@ -170,8 +173,49 @@ normalizeResponse(store, primaryModelClass, payload, id, requestType)
 modelFor(modelName)
 ```
 
-We would deprecate the class argument being passed in, and instruct the user to refactor towards using the Snapshot or the schema information. For backwards compatibility, we would still lookup the class from the registry, and if we found a class we would return it with a deprecation, but would no longer error if null was returned. We would deprecate accessing the class when passed to serializer/adapter by wrapping it in a proxy in dev mode
+### Serializing
+
+When serializing we already have a `Snapshot` passed in as an argument which has all of the information that a `ModelClass` provides. We would deprecate the class argument being passed in, and instruct the user to refactor towards using the `Snapshot`.
+
+For example: 
+```ts
+createRecord(store, type, snapshot) {
+  let url = `/api/${type.modelName}`;
+}
+```
+
+would become
+
+```ts
+createRecord(store, type, snapshot) {
+  let url = `/api/${snapshot.modelName}`;
+}
+```
+
+ For backwards compatibility, we would still lookup the class from the registry, and if we found a class we would return it with a deprecation, but would no longer error if null was returned. If we did not find the class, we would create a shim class that exposed a deprecated `modelName`. We would deprecate accessing the class when passed to serializer/adapter by wrapping it in a proxy in dev mode. Currently Snapshots contain all of the data that is available on the Model class which would be needed for serializing/normalizing.
+
+### Normalizing
+
+While `Snapshot` roughly corresponds to a `Request` like object to be used while serializing, and we intend to in the future refactor it to be a request object, we don't have a similar construct when normalizing. In the future, normalizing will have a corresponding `Response` like object exposed, but for the time being we can still deprecate using the `primaryModelClass` argument for anything other than `modelName`. If a user accesses any other property, like for example `eachAttribute` on the `primaryModelClass` they will
+receive a deprecation, to refactor using the public store methods. For example:
+
+```ts
+normalizeResponse(store, primaryModelClass, payload, id, requestType) {
+  primaryModelClass.eachAttribute((attr) => ...do something)
+}
+```
+
+would become
+
+```ts
+normalizeResponse(store, primaryModelClass, payload, id, requestType) {
+   let attrs = store.attributesDefitionFor(primaryModelClass.modelName);
+   Object.keys(attrs).forEach((attr) => ...do something);
+}
+```
 
+If there was a custom model instance provided, and we had no corresponding class to pass in to `normalizeResponse` we would pass in a shim class that only exposed `modelName` property, thus allowing the normalization layers to continue working with custom model classes.
+ 
 ## How we teach this
```

Comment:
I'm a bit concerned about the learning story as we push folks to use Snapshots more and more. Currently they are marked as `private` API and there is no public way to generate one (for doing things like testing a serializer/adapter). We may want to add a public `store.createSnapshot` API and work with the learning team on how to properly document classes that are non-user-instantiated but are public API.

Reply:
:+1: on documenting, but don't think we need to expose a creation method in order to document. I'm happy to have better snapshot docs as part of this rfc